File: frontend/android/app/src/main/java/kim/sunho/goranireader/models/EventLog.kt
package kim.sunho.goranireader.models

import io.realm.RealmObject
import io.realm.annotations.PrimaryKey
import io.realm.annotations.Required
import kotlinx.serialization.Serializable
import java.util.*

open class EventLog (
    @PrimaryKey
    @Required
    var id: String = "",
    var time: String = "",
    var type: String = "",
    var payload: String = ""
) : RealmObject()

@Serializable
data class ELPaginatePayload (
    val bookId: String = "",
    val chapterId: String = "",
    val time: Int = 0,
    val sids: List<String> = ArrayList(),
    val wordUnknowns: List<PaginateWordUnknown> = ArrayList(),
    val sentenceUnknowns: List<PaginateSentenceUnknown> = ArrayList()
)

@Serializable
data class ELUnknownWord (
    val bookId: String = "",
    val chapterId: String = "",
    val sentenceId: String = "",
    val wordIndex: Int = 0,
    val word: String = "",
    val def: String = ""
)

@Serializable
data class ELUnknownSentence (
    val bookId: String = "",
    val chapterId: String = "",
    val sentenceId: String = ""
)

File: frontend/ios/App/Models/EventLog.swift
//
// Copyright © 2019 Sunho Kim. All rights reserved.
//

import Foundation
import RealmSwift

class EventLog: Object, Codable {
    @objc dynamic var id: String = ""
    @objc dynamic var type: String = ""
    @objc dynamic var time: Date = Date()
    @objc dynamic var payload: String = ""
    override static func primaryKey() -> String? {
        return "id"
    }
}

struct EventLogMap: Codable {
    let type: String
    let time: Date
    let payload: String
}

protocol EventLogPayload: Codable {
    static func type() -> String
}

struct ELPaginatePayload: EventLogPayload {
    let bookId: String
    let chapterId: String
    let time: Int
    let sids: [String]
    let wordUnknowns: [PaginateWordUnknown]
    let sentenceUnknowns: [PaginateSentenceUnknown]
    
    static func type() -> String {
        return "paginate"
    }
}

struct ELUnknownSentencePayload: EventLogPayload {
    let bookId: String
    let chapterId: String
    let sentenceId: String
    
    static func type() -> String {
        return "unknown_sentence"
    }
}

struct ELSubmitQuestionPayload: EventLogPayload {
    let bookId: String
    let chapterId: String
    let questionId: String
    let option: String
    let right: Bool
    let time: Int
    
    static func type() -> String {
        return "submit_question"
    }
}

struct PaginateWordUnknown: Codable {
    let sentenceId: String
    let word: String
    let wordIndex: Int
    let time: Int
}

struct PaginateSentenceUnknown: Codable {
    let sentenceId: String
    let time: Int
}
